### ISSUE TITLE
bitcoind sometimes segfaulting on shutdown

### ISSUE THREAD

**Issue by c0deright** [bitcoin-sv/bitcoin-sv#102]
### Describe the issue

Since v0.2.0 we observe unclean shutdowns due to segfaulting bitcoind which results in BerkeleyDB logfiles being present in `<datadir>/database/` directory causing redownload of ~300 blocks on next startup.

Just now I upgraded several nodes from v0.2.0 to v0.2.1. On several of them shutdown of v0.2.0 was unclean due to segfault.

Retested with v0.2.1 on one of the nodes by stopping bitcoind where again during shutdown of v0.2.1 a segfault occured.

Starting/shutdown after that did result in a clean shutdown (no segfault).

### Can you reliably reproduce the issue?

No. Sometimes it segfaults during shutdown, sometimes it doesn't.

#### If so, please list the steps to reproduce below:
1. start bitcoind
2. wait until RPC responds
3. shutdown bitcoind

### Expected behaviour

debug.log of a clean shutdown:

```
2019-10-21 12:43:32 tor: Thread interrupt
2019-10-21 12:43:32 addcon thread exit
2019-10-21 12:43:32 scheduler thread interrupt
2019-10-21 12:43:32 torcontrol thread exit
2019-10-21 12:43:32 Preload wait for shutdown done
2019-10-21 12:43:32 Checking Thread shutdown
2019-10-21 12:43:32 Shutdown: In progress...
2019-10-21 12:43:32 net thread exit
2019-10-21 12:43:33 msghand thread exit
2019-10-21 12:43:35 opencon thread exit
2019-10-21 12:43:35 ConnmanPool ThreadPool thread 1 stopping
2019-10-21 12:43:35 ConnmanPool ThreadPool thread 0 stopping
2019-10-21 12:43:35 Dumped mempool: 8.6e-05s to copy, 0.003493s to dump
2019-10-21 12:43:36 Shutdown: done
```

No segfault logged

### Actual behaviour

debug.log:

```
2019-10-21 12:41:36 tor: Thread interrupt
2019-10-21 12:41:36 torcontrol thread exit
2019-10-21 12:41:36 addcon thread exit
2019-10-21 12:41:36 net thread exit
2019-10-21 12:41:36 opencon thread exit
2019-10-21 12:41:36 scheduler thread interrupt
2019-10-21 12:41:36 Preload wait for shutdown done
2019-10-21 12:41:36 Checking Thread shutdown
2019-10-21 12:41:36 Shutdown: In progress...
2019-10-21 12:41:36 msghand thread exit
2019-10-21 12:42:20
```

Notice the incomplete shutdown log here.

Kernel log:

```
kernel: [255027.335114] bitcoind[22442]: segfault at 50388 ip 0000561a59fe46dc sp 00007fd4ffffed70 error 4 in bitcoind[561a59e5b000+541000]
```

### What version of bitcoin-sv are you using?
v0.2.1 selfcompiled on Ubuntu Xenial with GCC-7 runtime packages installed from https://launchpad.net/~ubuntu-toolchain-r/+archive/ubuntu/test:

```
gcc-9-base_9.1.0-2ubuntu2~16.04_amd64.deb
libgcc1_1:9.1.0-2ubuntu2~16.04_amd64.deb
libstdc++6_9.1.0-2ubuntu2~16.04_amd64.deb
```

### Machine specs:
- OS: Ubuntu 16.04 LTS + above packages

### Any extra information that might be useful in the debugging process.

v1.x did not show this behaviour.

#### compile flags

CC=gcc-7 CXX=g++-7 ./configure --with-incompatible-bdb --disable-tests --disable-bench && make -j $(nproc)

#### ldd

```
% ldd bitcoind
        linux-vdso.so.1 =>  (0x00007ffd1bffd000)
        libboost_system.so.1.58.0 => /usr/lib/x86_64-linux-gnu/libboost_system.so.1.58.0 (0x00007f03ea3d6000)
        libboost_filesystem.so.1.58.0 => /usr/lib/x86_64-linux-gnu/libboost_filesystem.so.1.58.0 (0x00007f03ea1be000)
        libboost_program_options.so.1.58.0 => /usr/lib/x86_64-linux-gnu/libboost_program_options.so.1.58.0 (0x00007f03e9f40000)
        libboost_thread.so.1.58.0 => /usr/lib/x86_64-linux-gnu/libboost_thread.so.1.58.0 (0x00007f03e9d1a000)
        libboost_chrono.so.1.58.0 => /usr/lib/x86_64-linux-gnu/libboost_chrono.so.1.58.0 (0x00007f03e9b12000)
        libdb_cxx-5.3.so => /usr/lib/x86_64-linux-gnu/libdb_cxx-5.3.so (0x00007f03e973d000)
        libcrypto.so.1.0.0 => /lib/x86_64-linux-gnu/libcrypto.so.1.0.0 (0x00007f03e92f8000)
        libevent_pthreads-2.0.so.5 => /usr/lib/x86_64-linux-gnu/libevent_pthreads-2.0.so.5 (0x00007f03e90f5000)
        libevent-2.0.so.5 => /usr/lib/x86_64-linux-gnu/libevent-2.0.so.5 (0x00007f03e8eaf000)
        libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f03e8acc000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f03e87c3000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f03e85ab000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f03e838e000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f03e7fc4000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f03e7dbc000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f03e7bb8000)
        libevent_core-2.0.so.5 => /usr/lib/x86_64-linux-gnu/libevent_core-2.0.so.5 (0x00007f03e798e000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f03ead32000)
```

**Comment by Danconnolly** [comment]
Thank you for your report. This was fixed in version 0.2.2.beta here https://github.com/bitcoin-sv/bitcoin-sv/commit/74a6f61e28a7816f2e503c76b1dfeaeef49cdbaf

**Comment by c0deright** [comment]
Thanks.
